### PR TITLE
add kube-proxy mode in kubeadm config file

### PR DESCRIPTION
**Pull request by Lion-Wei**
added kube-proxy mode in kubeadm config file, so that we can easily chose proxy mode when build cluster, espically if we want use ipvs mode. 
Already added kube-proxy-mode args in kubeadm, in [this pr](https://github.com/kubernetes/kubernetes/pull/53962) : https://github.com/kubernetes/kubernetes/pull/53962

### CODE REVIEW

**Comment by Lion-Wei** [comment]
//cc @divyenpatel @shashidharatd

**Comment by shashidharatd** [comment]
The configuration is missing in Kconfig, and also how the config is used to pass onto kubeadm.
@Lion-Wei, Please consider testing it locally before submitting the PR. There is no pre-submit tests which validate the changes for this repo.

**Comment by Lion-Wei** [comment]
@shashidharatd ,yeah, I saw that. I'll need some time to test this pr. Thanks.

**Comment by Lion-Wei** [comment]
I already tested, `jce.jsonnet` added a fied `kube-proxy-mode: %(proxy_mode)`, "configure-vm-kubeadm.sh" can get proxy_mode info by  func "get_metamata", and write into kubeadm config file, so that we can pass proxy mode to kubeadm. @shashidharatd

**Comment by Lion-Wei** [comment]
The kubeadm already supported kube-proxy ipvs mode, see this pr: #53962
Hope this pr can be approved ASAP, so that we can use kubeadm build ipvs CI job. 
I already tested in my local environment. Hope someone can take a review, really appreciate it!
/cc @shashidharatd @luxas

**Comment by Lion-Wei** [comment]
//cc @divyenpatel @shashidharatd   kindly cc.  : )

File: phase2/Kconfig [hunk 1]
```diff
@@ -15,6 +15,14 @@ config phase2.provider
 
 	  Valid options are (ignition, kubeadm).
 
+config phase2.proxy_mode
+	string "kube-proxy mode: userspace, iptables or ipvs"
```
Comment (pipejakob) [inline phase2/Kconfig]: I'm not very familiar with kube-proxy modes, but this configuration seems to imply that kubernetes-anywhere is supporting three different modes for the proxy. Looking at how this field is used, it looks like we only honor a binary decision: ipvs or non-ipvs. I don't know what mode kubeadm uses by default if you don't enable ipvs, but it seems like we should only advertise those two values here (whether that's userspace or iptables, I'm not sure).

Comment (Lion-Wei) [reply phase2/Kconfig]: Thanks for replaying.
Yeah, it was suppose to support those three modes, but for some reason kubeadm only support iptables and ipvs for now. Use iptables by default and use ipvs if we set feature gate `SupportIPVSProxyMode=true`. 
This is short-term solution, and they were planing use kube-proxy ComponentConfig in kubeadm, which will definitely solve this problem.
There are more information in this pr : kubernetes/kubernetes#53962

Comment (pipejakob) [reply phase2/Kconfig]: Sure, in that case I would just remove `userspace` from the list for now so it doesn't mislead users.

Comment (Lion-Wei) [reply phase2/Kconfig]: Thanks, should I do that?

Comment (pipejakob) [reply phase2/Kconfig]: Oh, just here in the KConfig file. These values aren't enforced by the configuration system, but are used as user-prompted documentation as someone sets up their configuration. If you want to see it in practice, you can step into our Docker dev container with `make docker-dev` and then run `make menuconfig` inside it to use the curses-based frontend. Your configuration from KConfig is shown to the user when they set these fields.

TL;DR: just change your KConfig lines 19 and 24. Right now, they list all three options, but should only list the two that are supported right now.

Comment (Lion-Wei) [reply phase2/Kconfig]: Thanks!

**Comment by pipejakob** [comment]
@Lion-Wei Could you take care of your merge conflict and rebase?

**Comment by Lion-Wei** [comment]
@pipejakob Sure, rebased, thanks for reminding.

File: phase1/gce/gce.jsonnet [hunk 2]
```diff
@@ -185,6 +186,7 @@ function(cfg)
             "k8s-kubeadm-master-ip": "${google_compute_instance.%(master_instance)s.network_interface.0.address}" % names,
           } + if p2.provider == "kubeadm" then {
             "k8s-kubeadm-token": "${var.kubeadm_token}",
+            "kube-proxy-mode": "%(proxy_mode)s" % p2,
```
Comment (pipejakob) [inline phase1/gce/gce.jsonnet]: Nitpick: This shouldn't be needed on the node template, right? It looks like only the master template uses it, so this will just get ignored here.

File: phase1/gce/configure-vm.sh [hunk 3]
```diff
@@ -20,6 +20,7 @@ KUBEADM_ENABLE_CLOUD_PROVIDER=$(get_metadata "k8s-kubeadm-enable-cloud-provider"
 KUBEADM_ADVERTISE_ADDRESSES=$(get_metadata "k8s-kubeadm-advertise-addresses")
 KUBEADM_CNI_PLUGIN=$(get_metadata "k8s-kubeadm-cni-plugin")
 KUBEADM_MASTER_IP=$(get_metadata "k8s-kubeadm-master-ip")
+KUBEPROXY_MODE=$(get_metadata "kube-proxy-mode")
```
Comment (pipejakob) [inline phase1/gce/configure-vm.sh]: Nitpick, but for consistency can you name this `k8s-kube-proxy-mode` or something similar with a `k8s-` prefix? It just makes it easier to find all of the metadata related to Kubernetes. This won't affect user configuration, so I don't mind if you want to merge first for testing and then rename this in a future PR.

File: phase1/gce/gce.jsonnet (hunk 2 above)
Comment (Lion-Wei) [reply phase1/gce/gce.jsonnet]: Yes, my mistake, I have fixed that, thanks.

File: phase2/kubeadm/configure-vm-kubeadm-master.sh [hunk 4]
```diff
@@ -33,4 +33,11 @@ cloudProvider: "$CLOUD_PROVIDER"
 EOF
 fi
 
+if [[ "${KUBEPROXY_MODE}" == "ipvs" ]]; then
```
Comment (shashidharatd) [inline phase2/kubeadm/configure-vm-kubeadm-master.sh]: We need this variable `KUBEPROXY_MODE` set from other phase1 providers as well.
Can you please add this variable in openstack provider as well.

Comment (Lion-Wei) [reply phase2/kubeadm/configure-vm-kubeadm-master.sh]: Sure, I added in openstack,

File: phase1/openstack/configure-vm.sh [hunk 5]
```diff
@@ -13,6 +13,7 @@ KUBEADM_ENABLE_CLOUD_PROVIDER="${k8s_kubeadm_enable_cloud_provider}"
 KUBEADM_ADVERTISE_ADDRESSES="${k8s_kubeadm_advertise_addresses}"
 KUBEADM_CNI_PLUGIN="${k8s_kubeadm_cni_plugin}"
 KUBEADM_MASTER_IP="${k8s_kubeadm_master_ip}"
+KUBEPROXY_MODE=$(get_metadata "k8s_kubeproxy_mode")
```
Comment (shashidharatd) [inline phase1/openstack/configure-vm.sh]: @Lion-Wei, we do not have get_metatdata function in openstack. please check above lines on how to assign the variable.

File: phase2/kubeadm/configure-vm-kubeadm-master.sh (hunk 4 above)
Comment (shashidharatd) [reply phase2/kubeadm/configure-vm-kubeadm-master.sh]: we cannot use the variable with curly braces like "${KUBEPROXY_MODE}", as it will conflict with terraform and/or jsonnet templating. See the other variables in the file and follow the similar pattern

File: phase1/openstack/configure-vm.sh (hunk 5 above)
Comment (Lion-Wei) [reply phase1/openstack/configure-vm.sh]: Sorry for that... : (

**Comment by Lion-Wei** [comment]
Now kubeadm support kube-proxy component config, so I changed kubeadm config file kube-proxy part to : 
```
kubeProxy:
  featureGates: SupportIPVSProxyMode=true
  mode: "$KUBEPROXY_MODE"
```
/cc @shashidharatd @pipejakob

**Comment by shashidharatd** [comment]
/lgtm
Thanks @Lion-Wei !

**Comment by luxas** [comment]
Or due to the time requirements, I'll merge this, it has already been "nearly" LGTM'd by Jakob.
@shashidharatd also LGTM'd it so let's go :)

**Comment by Lion-Wei** [comment]
Thanks everyone. @shashidharatd @pipejakob @luxas
